### ISSUE TITLE
Potential open telemetry memory leak in Spring Boot WebFlux service

### ISSUE THREAD

**Issue by zoobert** [open-telemetry/opentelemetry-java#5546]
**Describe the bug**
When I instrument a very simple WebFlux Spring Boot app that simply front's a postgres database, I see a large amount of memory retained by `io.micrometer.observation.SimpleObservation` to the point where with large amounts of requests we start to receive OOM exceptions.

**Steps to reproduce**
The following is a repo containing a slimmed-down version of the problem: https://github.com/zoobert/otel-memory. This repo contains both the server with the problem and a client which makes many calls to the server to insert records.

**What did you expect to see?**
I had hoped to use open telemetry in my Spring Boot application. Micrometer cannot be eliminated from the mix because I am using both Project Reactor and Spring Security use it. I'm also using Project Reactor's automatic context propagation to get the trace context to the Reactor Context and Kotlin's coroutines.

```
Hooks.enableAutomaticContextPropagation()
```

**What did you see instead?**
Within our full application, we are running into OOM exceptions when collecting open telemetry trace spans.

**What version and what artifacts are you using?**
Artifacts: io.opentelemetry:opentelemetry-exporter-otlp,io.micrometer:micrometer-tracing-bridge-otel
Version: 1.26.0
How did you reference these artifacts? build.gradle.kts

**Environment**
Compiler: Amazon Corretto JDK 19.0.2+7-FR
OS: MacOS Ventura 13.4

**Additional context**
The problem seems to be related to r2dbc pooling, Reactor content propagation, and/or spring security. If I disable any of these the problem goes away or is diminished.

**Comment by jkwatson** [comment]
It sounds like this is probably _not_ due to core otel java libraries, correct? If that's the case, we should probably move this issue over to the opentelemetry-java-instrumentation where the issue likely lies.

**Comment by zoobert** [comment]
Yes, sorry. I'll reenter it there unless there is a way to actually move it over without re-entering it.

**Comment by jkwatson** [comment]
I can move it.
